### PR TITLE
Multiple runs in a single evaluation

### LINKED ISSUE #108: Multiple Runs
Runs per all tasks, i.e. for models A,B,C execute ABCABC (*not* AABBCC).

- [x] add runs as CLI arg
- [x] do multiple runs
- [x] ensure the scores per runs are just summed up

### PR DESCRIPTION
**Pull request by bauersimon**
Closes #108

### CODE REVIEW

**Comment by Munsio** [comment]
The skipping of tests doesn't work but the rest of the code looks good imho.

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 1]
```diff
@@ -46,6 +46,8 @@ type Evaluate struct {
 	Repositories []string `long:"repository" description:"Evaluate with this repository. By default all repositories are used."`
 	// ResultPath holds the directory path where results should be written to.
 	ResultPath string `long:"result-path" description:"Directory path where results should be written to. The placeholder \"%datetime%\" can be used for the current date and time." default:"evaluation-%datetime%"`
+	// Runs is the number of runs to perform.
+	Runs uint `long:"runs" short:"n" description:"Number of runs to perform." default:"1"`
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:50]: No short options. It gets confusing fast.

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 2]
```diff
@@ -229,60 +231,72 @@ func (command *Evaluate) Execute(args []string) (err error) {
 	assessments := report.NewAssessmentPerModelPerLanguagePerRepository(maps.Values(modelsSelected), maps.Values(languagesSelected), command.Repositories)
 	problemsPerModel := map[string][]error{}
 	{
-		for _, languageID := range command.Languages {
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:232]: Option check: "runs" should not be 0

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 3]
```diff
@@ -229,60 +231,72 @@ func (command *Evaluate) Execute(args []string) (err error) {
...
 	}
 
 	// Evaluating models and languages.
 	log.Printf("Evaluating models and languages")
-	for _, languageID := range command.Languages {
-		languagePath := filepath.Join(command.TestdataPath, languageID)
-		repositories, err := os.ReadDir(languagePath)
-		if err != nil {
-			log.Panicf("ERROR: language path %q cannot be accessed: %s", languagePath, err)
+	for r := uint(0); r < command.Runs; r++ {
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:262]: meeeen the "hide whitespaces" option for the diff saves the day again.

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 4]
```diff
@@ -299,7 +313,7 @@ func (command *Evaluate) Execute(args []string) (err error) {
 		}
 	}
 	if isOnlyPlainRepositories {
-		totalScore = uint(len(languagesSelected))
+		totalScore = uint(len(languagesSelected)) * command.Runs
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:316]: Looking forward to fixing this with an "optimal run"

File: cmd/eval-dev-quality/cmd/evaluate_test.go [hunk 5]
```diff
@@ -162,7 +162,10 @@ func TestEvaluateExecute(t *testing.T) {
 				require.NoError(t, err)
 			}
 			sort.Strings(actualResultFiles)
-			expectedResultFiles := make([]string, 0, len(tc.ExpectedResultFiles))
+			var expectedResultFiles []string
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate_test.go:165]: Why is that change necessary?

File: cmd/eval-dev-quality/cmd/evaluate_test.go [hunk 6]
```diff
@@ -543,6 +546,77 @@ func TestEvaluateExecute(t *testing.T) {
 		})
 	})
 
+	t.Run("Runs", func(t *testing.T) {
+		validate(t, &testCase{
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate_test.go:550]: Drop this case. we should not test go-flags in our test suite.

File: cmd/eval-dev-quality/cmd/evaluate_test.go [hunk 7]
```diff
@@ -543,6 +546,77 @@ func TestEvaluateExecute(t *testing.T) {
...
+
+			ExpectedPanicContains: "Could not parse arguments",
+		})
+		validate(t, &testCase{
+			Name: "Multiple",
+
+			Arguments: []string{
+				"--model", "symflower/symbolic-execution",
+				"--repository", filepath.Join("golang", "plain"),
+				"--runs=3",
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate_test.go:573]: 2 is enough.

File: cmd/eval-dev-quality/cmd/evaluate_test.go [hunk 8]
```diff
@@ -543,6 +546,77 @@ func TestEvaluateExecute(t *testing.T) {
...
+					metrics.Assessments{
+						metrics.AssessmentKeyCoverageStatement: 30,
+						metrics.AssessmentKeyFilesExecuted:     3,
+						metrics.AssessmentKeyResponseNoError:   3,
+						metrics.AssessmentKeyResponseNoExcess:  3,
+						metrics.AssessmentKeyResponseWithCode:  3,
+					},
+				}, []uint{42})
+				assert.Greater(t, actualAssessments[0][metrics.AssessmentKeyProcessingTime], uint(0))
+				if !assert.Equal(t, 1, strings.Count(output, "Evaluation score for")) {
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate_test.go:587]: Drop that if statement. just put a 

defer func () ... if t.Failed() then log

into the validate function.

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 9]
```diff
@@ -46,6 +46,8 @@ type Evaluate struct {
 	Repositories []string `long:"repository" description:"Evaluate with this repository. By default all repositories are used."`
 	// ResultPath holds the directory path where results should be written to.
 	ResultPath string `long:"result-path" description:"Directory path where results should be written to. The placeholder \"%datetime%\" can be used for the current date and time." default:"evaluation-%datetime%"`
+	// Runs is the number of runs to perform.
```
Comment (zimmski) [inline cmd/eval-dev-quality/cmd/evaluate.go:49]: -> `Runs holds the number of runs to perform.`

File: cmd/eval-dev-quality/cmd/evaluate_test.go (hunk 5 above)
Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate_test.go:165]: Cause `osutil.FilesRecursive` returns a `nil` list for the actual files but we then call `make` with the length (0) but that is an empty list... not a `nil` list. Hence the assertion would fail.
